source/common/config/datasource.h

```diff
@@ -23,21 +24,87 @@ namespace DataSource {
  * Read contents of the DataSource.
  * @param source data source.
  * @param allow_empty return an empty string if no DataSource case is specified.
- * @param api reference to the Api object
+ * @param api reference to the Api.
  * @param max_size max size limit of file to read, default 0 means no limit, and if the file data
- * would exceed the limit, it will throw a EnvoyException.
+ * would exceed the limit, it will return an error status.
  * @return std::string with DataSource contents. or an error status if no DataSource case is
  * specified and !allow_empty.
  */
 absl::StatusOr<std::string> read(const envoy::config::core::v3::DataSource& source,
                                  bool allow_empty, Api::Api& api, uint64_t max_size = 0);
 
+/**
+ * Read contents of the file.
+ * @param path file path.
+ * @param api reference to the Api.
+ * @param allow_empty return an empty string if the file is empty.
+ * @param max_size max size limit of file to read, default 0 means no limit, and if the file data
+ * would exceed the limit, it will return an error status.
+ * @return std::string with file contents. or an error status if the file does not exist or
+ * cannot be read.
+ */
+absl::StatusOr<std::string> readFile(const std::string& path, Api::Api& api, bool allow_empty,
```

Comment:
High level question before I dig in.

so read() already handles  envoy::config::core::v3::DataSource::SpecifierCase::kFilename:
If we need a different one for watched files, why not have a kDynamicFile Datasource and have this handled as a case in read()? 

cc @adisuissa @envoyproxy/api-shepherds for guidance on dynamic configuration sources
/wait-any

Reply:
Dynamic file support require to use the DataSourceProvider as return type, then it need to change all use cases of current DataSource.

Reply:
How does this related to `WatchedDirectory` and more specifically (for a single filename), `PathConfiSource`?

At the very least, there should be some overlap in implementation, and some guidance on how we want to think about the API. The watched directory concept is tricky, as it related to TLS cert rotation, which involves multiple files, and admittedly is not really what we want for a single file. But, I think `PathConfigSource` is a lot closer to what is going on in this PR.

Reply:
@htuch Specific to this scenario, maybe we can add an additional `WatchedDirectory` to the API where require dynamic `DataSource`. (Or split the directory from the filename and watch the directory directly?)

But anyway, I am inclined don't to add a new field to the `DataSource` directly. A fact is that the `DataSource` as a basic API is widely used by various modules or extensions.

If we add new dynamic file field to the `DataSource`, then the end users will expect all the features that use the `DataSource` support the dynamic loading. But this actually depends on the feature implementations. There are lots of features never require the dynamic loading (or dynamic loading will make the feature too complex and doesn't deserve it.).

So, rather than extending the `DataSource` directly, I personally prefer to add new API at where require the dynamic `DataSource`.

cc @alyssawilk

Reply:
Yeah, I get that, and it makes sense. But I'd like to see better guidance on when we consume configuration via `DataSource` vs. `PathConfigSource`. The latter has some involved semantics that is intended to facilitate sensible behavior across symlink based cert _and_ configuration rotation. If you don't need this, then something much simpler like this approach works. But, I think we need to be very deliberate each time we decide to use a simple `DataSource` for `PathConfigSource` for watching, to make sure we don't end up creating something that doesn't work well under some fairly common pattern for config rotation (even Envoy uses this for runtime symlinks swaps).

Reply:
One potential point of confusion is SDS using a dynamic data source. Feels like it'd trigger twice.

Reply:
> But, I think we need to be very deliberate each time we decide to use a simple DataSource vs. PathConfigSource for watching, to make sure we don't end up creating something that doesn't work well under some fairly common pattern for config rotation (even Envoy uses this for runtime symlinks swaps).

Yeah. Basically, this PR is an enhancement to the DataSource to provide a helper to watch the filename of the DataSource.

So, this should only be used when DataSource is wanted and file watching is necessary. In any other ways (for example, at the case where just want to watch a file, doesn't need the `inline_string/inline_bytes` of DataSource), this should not be used.

I will document this point as part of the new helper.

Reply:
@htuch  Can you please take a look at this when you get chance?